library/src/main/java/com/google/android/exoplayer/extractor/ts/Mpeg2Reader.java

```diff
@@ -0,0 +1,189 @@
+package com.google.android.exoplayer.extractor.ts;
+
+import com.google.android.exoplayer.C;
+import com.google.android.exoplayer.MediaFormat;
+import com.google.android.exoplayer.extractor.TrackOutput;
+import com.google.android.exoplayer.util.Assertions;
+import com.google.android.exoplayer.util.MimeTypes;
+import com.google.android.exoplayer.util.ParsableByteArray;
+
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parses a continuous mpeg2 byte stream and extracts individual frames.
+ */
+/* package */ final class Mpeg2Reader extends ElementaryStreamReader {
+    private static final String TAG = "Mpeg2Reader";
+
+    private int flags;
+    private boolean seqHeaderFound;
+    private boolean frameCollected;
+    private boolean isSync;
+
+    private int width;
+    private int height;
+
+    private long timeUs;
+    private long samplePosition;
+
+    private boolean foundFirstSample;
+    private long totalBytesWritten;
+
+    public Mpeg2Reader(TrackOutput output) {
+        super(output);
+    }
+
+    @Override
+    public void seek() {
+        seqHeaderFound = false;
+        samplePosition = 0;
+        foundFirstSample = false;
+        totalBytesWritten = 0;
+    }
+
+    @Override
+    public void consume(ParsableByteArray data, long pesTimeUs, boolean startOfPacket) {
+        while (data.bytesLeft() > 0) {
+            int frameAtOffset = 0;
+            int offset = data.getPosition();
+            int limit = data.limit();
+            byte[] dataArray = data.data;
+            // Append the data to the buffer.
+            totalBytesWritten += data.bytesLeft();
+            output.sampleData(data, data.bytesLeft());
+
+            if (!seqHeaderFound && skipToNextSync(dataArray, offset, limit)) {
+                if (seqHeaderFound) {
+                    flags = C.SAMPLE_FLAG_SYNC;
+                    foundFirstSample = true;
+                    timeUs = pesTimeUs;
+                    findSampleSizeAndPrepareMediaFormat(dataArray, offset); // Seq header has been already been found
+                }
+            }
+
+            if (startOfPacket && seqHeaderFound) {
+
+                frameAtOffset = findValidStartCode(dataArray, offset, limit);
+
+                if (frameAtOffset < limit) {
+                    frameCollected = true;
+                    if (!foundFirstSample)
+                        flags = (isSync == true) ? C.SAMPLE_FLAG_SYNC : 0;
```

Comment:
Move this line to just before the output.sampleMetadata call, and make flags a local variable there.